File: 002/002.py
def solution(n=4000000):
    """Computes sum of all even-valued Fibonacci terms that are less than n."""
    a, b, ans = 1, 2, 0
    while b < n:
        if b & 1 == 0:
            ans += b
        a, b = b, a + b
    return ans


def solution_2(n=4000000):
    """Computes sum of all even-valued Fibonacci terms that are less than n."""
    a, b, ans = 1, 2, 0
    while b < n:
        ans += b
        for i in range(3):
            a, b = b, a + b
    return ans


def solution_3(n=4000000):
    """Computes sum of all even-valued Fibonacci terms that are less than n."""
    return sum(even_fib_gen(n))


def solution_4(n=4000000):
    """Computes sum of all even-valued Fibonacci terms that are less than n."""
    return sum(even_fib_gen_2(n))


def even_fib_gen(n):
    """Generator for even-valued Fibonacci terms that are less than n."""
    a, b = 1, 2
    while b < n:
        if b & 1 == 0:
            yield b
        a, b = b, a + b


def even_fib_gen_2(n):
    """Generator for even-valued Fibonacci terms that are less than n."""
    a, b = 1, 2
    while b < n:
        yield b
        for i in range(3):
            a, b = b, a + b


if __name__ == '__main__':
    from timeit import timeit
    from math import log10

    print('Answer: {}'.format(solution()))

    t_elapsed = timeit('solution()',
                       setup='from __main__ import solution',
                       number=3) / 3

    if t_elapsed < 3:
        run_time  = 10
        n_iters   = int(run_time / t_elapsed)
        t_elapsed = timeit('solution()',
                           setup='from __main__ import solution',
                           number=n_iters) / n_iters

    exp = log10(t_elapsed)

    if exp > 0:
        t_string = '{:.3g} s'.format(t_elapsed)
    elif exp > -3:
        t_string = '{:.3g} ms'.format(t_elapsed * 1e3)
    elif exp > -6:
        t_string = '{:.3g} µs'.format(t_elapsed * 1e6)
    else:
        t_string = '{:.3g} ns'.format(t_elapsed * 1e9)

    print('Average Duration: ' + t_string)

File: 002/002.cpp
#include <stdio.h>
#include <math.h>
#include <sys/time.h>

unsigned long solution(unsigned long n=4000000) {

    unsigned long a   = 1;
    unsigned long b   = 2;
    unsigned long tmp = 0;
    unsigned long ans = 0;

    while (b < n) {

        if (!(b & 1)) {
            ans += b;
        }

        tmp = a;
        a   = b;
        b  += tmp;

    }

    return ans;
}

unsigned long solution_2(unsigned long n=4000000) {

    unsigned long a   = 1;
    unsigned long b   = 2;
    unsigned long tmp = 0;
    unsigned long ans = 0;

    while (b < n) {

        ans += b;

        for (short i = 0; i < 3; i++) {
            tmp = a;
            a   = b;
            b  += tmp;
        }

    }

    return ans;
}

//////////////////////////////////////////////////

typedef unsigned long long timestamp_t;

static timestamp_t

get_timestamp () {
  struct timeval now;
  gettimeofday (&now, NULL);
  return  now.tv_usec + (timestamp_t)now.tv_sec * 1000000;
}

int main() {

    printf("Answer: %lu\n", solution());

    timestamp_t t0 = get_timestamp();

    for (int i = 0; i < 3; ++i) {
        solution();
    }

    timestamp_t t1 = get_timestamp();

    double t_elapsed = (t1 - t0) / 1000000.0L / 3;

    if (t_elapsed < 3) {
         int run_time = 10;
         unsigned long long n_iters = run_time / t_elapsed;

         timestamp_t t0 = get_timestamp();

         for (int i = 0; i < n_iters; ++i) {
             solution();
         }

         timestamp_t t1 = get_timestamp();
         t_elapsed = (t1 - t0) / 1000000.0L / n_iters;
    }

    double exp_t = log10(t_elapsed);

    if (exp_t > 0) {
        printf("Average Duration: %.3g s\n", t_elapsed);
    } else if (exp_t > -3) {
        printf("Average Duration: %.3Lg ms\n", t_elapsed * 1000.0L);
    } else if (exp_t > -6) {
        printf("Average Duration: %.3Lg µs\n", t_elapsed * 1000000.0L);
    } else {
        printf("Average Duration: %.3Lg ns\n", t_elapsed * 1000000000.0L);
    }

    return 0;
}
